### DISCUSSION TITLE
How to chain commands in CLI?

### DISCUSSION THREAD

**Discussion by ivanjaros** [libvips/libvips#4256] [Q&A] [answered]
From what I have read, it seems there is no way to chain transformations in CLI.
So I am running multiple commands one after another.
The problem I have is that if src and dst files are the same file, vips crashes with write error.
Is there a way to transform existing image without creating a new one?
I've seen mention of `memory=true` in documentation, which should load the entire file into memory and avoid the temporary files "proxy". Is that the way to go or is there another option?

**Comment by jcupitt** [comment, marked as answer]
Hello @ivanjaros,

That's right, libvips supports streaming with coordinate transformations, so you can't use pipes to connect operations. You need something like python / ruby / php / etc. to chain operations together.

If you run something like:

```
vips flip a.png b.png horizontal
```

libvips will run the PNG decode and encode in parallel, reading scanlines from `a.png` and writing to `b.png`. 

This means you can operate on images of any size (only a few pixels are ever kept in memory at once), and with good speed (PNG encode and decode is slow, so this gets you an easy x2 speedup). The downside is that you can't read and write from the same file.

I usually do something like this in scripts:

```bash
# transpose image
$tmp=tmp-$$
mkdir $tmp
vips flip $1 $tmp/t1.v horizontal
vips flip $tmp/t1.v $2 vertical
rm -rf $tmp
```

- under bash, `$$` expands to the shell process id, so `tmp-$$` becomes something like `tmp-80598` ... it's an easy way to make a unique number
- I run a series of commands using vips format files in `$tmp` as intermediates ... vips format is very quick and preserves all metadata (files are large, on the downside)
- at the end, I remove the `$tmp` directory, and this will remove any left-over intermediates

**Reply by ivanjaros** [reply]
i can use that. thanks.

**Comment by ILoveGoulash** [comment]
You can look at the solution I reached in https://github.com/libvips/libvips/discussions/3443 if that's your thing. Here's the final script: https://git.sr.ht/~q3cpma/scripts/tree/master/item/vips.py (ignore all the "dynamic import" business).
